lib/pure/httpclient.nim

```diff
@@ -399,6 +399,58 @@ proc request*(url: string, httpMethod: string, extraHeaders = "",
   var hostUrl = if proxy == nil: r else: parseUri(url)
   var headers = substr(httpMethod, len("http"))
   # TODO: Use generateHeaders further down once it supports proxies.
+  
+  var s = newSocket()
+  if s == nil: raiseOSError(osLastError())
+  var port = net.Port(80)
+  if r.scheme == "https":
+    when defined(ssl):
+      sslContext.wrapSocket(s)
+      port = net.Port(443)
+    else:
+      raise newException(HttpRequestError,
+                "SSL support is not available. Cannot connect over SSL.")
+  if r.port != "":
+    port = net.Port(r.port.parseInt)
+
+
+  # get the socket ready. If we are connecting through a proxy to SSL,
+  # send the appropiate CONNECT header. If not, simply connect to the proper
+  # host (which may still be the proxy, for normal HTTP)
+  if proxy != nil and hostUrl.scheme == "https":
+    when defined(ssl):
+      var connectHeaders = "CONNECT "
```

Comment:
To be clear, the client sends an HTTP `CONNECT` request to the proxy, which opens a socket to the server. Then the client sends a normal TLS handshake request to the proxy, which forwards it unmodified to the server.
```
client --(HTTP CONNECT)--> proxy
proxy --(raw socket)-->server
client --(TLS handshake)--> proxy
proxy --(forward raw bytes of handshake)--> sever
server --(TLS response) --> proxy
proxy --(raw TLS response)--> client
```